File: hdf-diff/test/exit-codes.test.ts
import { describe, it, expect } from 'vitest';
import { computeExitCode, computeDetailedExitCode,
  EXIT_IDENTICAL, EXIT_DIFFERENCES, EXIT_ERROR,
  EXIT_DETAILED_IDENTICAL, EXIT_DETAILED_ERROR,
  EXIT_DETAILED_FIXES_ONLY, EXIT_DETAILED_REGRESSIONS_ONLY,
  EXIT_DETAILED_MIXED, EXIT_DETAILED_BASELINE_CHANGED,
  EXIT_DETAILED_DRIFT_ONLY,
} from '../src/exit-codes.js';
import type { ComparisonSummary } from '../src/types.js';

/**
 * Build a ComparisonSummary with sensible defaults (all unchanged).
 * Override individual fields as needed for each test scenario.
 */
function makeSummary(overrides: Partial<ComparisonSummary> = {}): ComparisonSummary {
  return {
    total: 10,
    fixed: 0,
    regressed: 0,
    new: 0,
    absent: 0,
    unchanged: 10,
    updated: 0,
    matchedCount: 10,
    unmatchedOldCount: 0,
    unmatchedNewCount: 0,
    ...overrides,
  };
}

// ── Constants ───────────────────────────────────────────────────────────

describe('exit code constants', () => {
  it('should export GNU diff compatible constants', () => {
    expect(EXIT_IDENTICAL).toBe(0);
    expect(EXIT_DIFFERENCES).toBe(1);
    expect(EXIT_ERROR).toBe(2);
  });

  it('should export detailed exit code constants', () => {
    expect(EXIT_DETAILED_IDENTICAL).toBe(0);
    expect(EXIT_DETAILED_ERROR).toBe(1);
    expect(EXIT_DETAILED_FIXES_ONLY).toBe(10);
    expect(EXIT_DETAILED_REGRESSIONS_ONLY).toBe(11);
    expect(EXIT_DETAILED_MIXED).toBe(12);
    expect(EXIT_DETAILED_BASELINE_CHANGED).toBe(13);
    expect(EXIT_DETAILED_DRIFT_ONLY).toBe(14);
  });
});

// ── computeExitCode (basic/GNU diff mode) ───────────────────────────────

describe('computeExitCode', () => {
  it('should return 0 when all requirements are unchanged (identical)', () => {
    const summary = makeSummary();
    expect(computeExitCode(summary)).toBe(0);
  });

  it('should return 1 when fixes exist', () => {
    const summary = makeSummary({ fixed: 3, unchanged: 7 });
    expect(computeExitCode(summary)).toBe(1);
  });

  it('should return 1 when regressions exist', () => {
    const summary = makeSummary({ regressed: 2, unchanged: 8 });
    expect(computeExitCode(summary)).toBe(1);
  });

  it('should return 1 when mixed fixes and regressions exist', () => {
    const summary = makeSummary({ fixed: 2, regressed: 1, unchanged: 7 });
    expect(computeExitCode(summary)).toBe(1);
  });

  it('should return 1 when only new controls exist', () => {
    const summary = makeSummary({
      new: 2, unchanged: 10, total: 12,
      unmatchedNewCount: 2, matchedCount: 10,
    });
    expect(computeExitCode(summary)).toBe(1);
  });

  it('should return 1 when only absent controls exist', () => {
    const summary = makeSummary({
      absent: 3, unchanged: 7, total: 10,
      unmatchedOldCount: 3, matchedCount: 7,
    });
    expect(computeExitCode(summary)).toBe(1);
  });

  it('should return 1 when new AND absent controls exist', () => {
    const summary = makeSummary({
      new: 1, absent: 1, unchanged: 8, total: 10,
      unmatchedNewCount: 1, unmatchedOldCount: 1, matchedCount: 8,
    });
    expect(computeExitCode(summary)).toBe(1);
  });

  it('should return 1 when only metadata drift (updated) exists', () => {
    const summary = makeSummary({ updated: 1, unchanged: 9 });
    expect(computeExitCode(summary)).toBe(1);
  });

  it('should return 0 for an empty comparison (total=0)', () => {
    const summary = makeSummary({
      total: 0, unchanged: 0, matchedCount: 0,
    });
    expect(computeExitCode(summary)).toBe(0);
  });
});

// ── computeDetailedExitCode ─────────────────────────────────────────────

describe('computeDetailedExitCode', () => {
  // Scenario 1: All unchanged → 0
  it('should return 0 when all requirements are unchanged (identical)', () => {
    const summary = makeSummary();
    expect(computeDetailedExitCode(summary)).toBe(0);
  });

  // Scenario 2: Fixes only → 10
  it('should return 10 when only fixes are present', () => {
    const summary = makeSummary({ fixed: 3, unchanged: 7 });
    expect(computeDetailedExitCode(summary)).toBe(10);
  });

  // Scenario 3: Regressions only → 11
  it('should return 11 when only regressions are present', () => {
    const summary = makeSummary({ regressed: 2, unchanged: 8 });
    expect(computeDetailedExitCode(summary)).toBe(11);
  });

  // Scenario 4: Mixed fixes and regressions → 12
  it('should return 12 when both fixes and regressions are present', () => {
    const summary = makeSummary({ fixed: 2, regressed: 1, unchanged: 7 });
    expect(computeDetailedExitCode(summary)).toBe(12);
  });

  // Scenario 5: Only new controls → 13
  it('should return 13 when only new controls are present', () => {
    const summary = makeSummary({
      new: 2, unchanged: 10, total: 12,
      unmatchedNewCount: 2, matchedCount: 10,
    });
    expect(computeDetailedExitCode(summary)).toBe(13);
  });

  // Scenario 6: Only absent controls → 13
  it('should return 13 when only absent controls are present', () => {
    const summary = makeSummary({
      absent: 3, unchanged: 7, total: 10,
      unmatchedOldCount: 3, matchedCount: 7,
    });
    expect(computeDetailedExitCode(summary)).toBe(13);
  });

  // Scenario 7: New AND absent (but no fixes/regressions) → 13
  it('should return 13 when both new and absent controls are present but no status changes', () => {
    const summary = makeSummary({
      new: 1, absent: 1, unchanged: 8, total: 10,
      unmatchedNewCount: 1, unmatchedOldCount: 1, matchedCount: 8,
    });
    expect(computeDetailedExitCode(summary)).toBe(13);
  });

  // Scenario 8: Only metadata drift (updated but no fixes/regressions/new/absent) → 14
  it('should return 14 when only metadata drift (updated) is present', () => {
    const summary = makeSummary({ updated: 1, unchanged: 9 });
    expect(computeDetailedExitCode(summary)).toBe(14);
  });

  // Scenario 9: Fixes + new controls → 10 (fixes take priority over baseline changes)
  it('should return 10 when fixes and new controls are present (fixes take priority)', () => {
    const summary = makeSummary({
      fixed: 2, new: 1, unchanged: 7, total: 10,
      unmatchedNewCount: 1, matchedCount: 9,
    });
    expect(computeDetailedExitCode(summary)).toBe(10);
  });

  // Scenario 10: Regressions + absent controls → 11 (regressions take priority)
  it('should return 11 when regressions and absent controls are present (regressions take priority)', () => {
    const summary = makeSummary({
      regressed: 1, absent: 2, unchanged: 7, total: 10,
      unmatchedOldCount: 2, matchedCount: 8,
    });
    expect(computeDetailedExitCode(summary)).toBe(11);
  });

  // Edge: empty comparison → 0
  it('should return 0 for an empty comparison (total=0)', () => {
    const summary = makeSummary({
      total: 0, unchanged: 0, matchedCount: 0,
    });
    expect(computeDetailedExitCode(summary)).toBe(0);
  });

  // Edge: fixes + regressions + new + absent → 12 (mixed takes priority)
  it('should return 12 when fixes, regressions, new, and absent all present', () => {
    const summary = makeSummary({
      fixed: 1, regressed: 1, new: 1, absent: 1, unchanged: 6, total: 10,
      unmatchedNewCount: 1, unmatchedOldCount: 1, matchedCount: 8,
    });
    expect(computeDetailedExitCode(summary)).toBe(12);
  });

  // Edge: updated + new → 13 (baseline changes take priority over drift)
  it('should return 13 when updated and new controls present but no status changes', () => {
    const summary = makeSummary({
      updated: 1, new: 1, unchanged: 8, total: 10,
      unmatchedNewCount: 1, matchedCount: 9,
    });
    expect(computeDetailedExitCode(summary)).toBe(13);
  });
});

File: hdf-diff/go/exitcodes_test.go
package diff

import (
	"testing"
)

// makeSummary builds a ComparisonSummary with sensible defaults (all unchanged).
// Override individual fields as needed for each test scenario.
func makeSummary(overrides ...func(*ComparisonSummary)) ComparisonSummary {
	s := ComparisonSummary{
		Total:             10,
		Fixed:             0,
		Regressed:         0,
		New:               0,
		Absent:            0,
		Unchanged:         10,
		Updated:           0,
		MatchedCount:      10,
		UnmatchedOldCount: 0,
		UnmatchedNewCount: 0,
	}
	for _, o := range overrides {
		o(&s)
	}
	return s
}

// ── Constants ───────────────────────────────────────────────────────────

func TestConstants_BasicExitCodes(t *testing.T) {
	t.Run("GNU diff compatible constants have correct values", func(t *testing.T) {
		if Identical != 0 {
			t.Errorf("Identical = %d, want 0", Identical)
		}
		if Differences != 1 {
			t.Errorf("Differences = %d, want 1", Differences)
		}
		if Error != 2 {
			t.Errorf("Error = %d, want 2", Error)
		}
	})
}

func TestConstants_DetailedExitCodes(t *testing.T) {
	t.Run("detailed exit code constants have correct values", func(t *testing.T) {
		if FixesOnly != 10 {
			t.Errorf("FixesOnly = %d, want 10", FixesOnly)
		}
		if RegressionsOnly != 11 {
			t.Errorf("RegressionsOnly = %d, want 11", RegressionsOnly)
		}
		if Mixed != 12 {
			t.Errorf("Mixed = %d, want 12", Mixed)
		}
		if BaselineChanged != 13 {
			t.Errorf("BaselineChanged = %d, want 13", BaselineChanged)
		}
		if DriftOnly != 14 {
			t.Errorf("DriftOnly = %d, want 14", DriftOnly)
		}
	})
}

// ── ComputeBasicExitCode ────────────────────────────────────────────────

func TestComputeBasicExitCode(t *testing.T) {
	tests := []struct {
		name     string
		summary  ComparisonSummary
		expected int
	}{
		{
			name:     "all unchanged returns 0 (identical)",
			summary:  makeSummary(),
			expected: 0,
		},
		{
			name: "some fixed returns 1",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Fixed = 3
				s.Unchanged = 7
			}),
			expected: 1,
		},
		{
			name: "some regressed returns 1",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Regressed = 2
				s.Unchanged = 8
			}),
			expected: 1,
		},
		{
			name: "mixed fixes and regressions returns 1",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Fixed = 2
				s.Regressed = 1
				s.Unchanged = 7
			}),
			expected: 1,
		},
		{
			name: "only new controls returns 1",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.New = 2
				s.Unchanged = 10
				s.Total = 12
				s.UnmatchedNewCount = 2
				s.MatchedCount = 10
			}),
			expected: 1,
		},
		{
			name: "only absent controls returns 1",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Absent = 3
				s.Unchanged = 7
				s.Total = 10
				s.UnmatchedOldCount = 3
				s.MatchedCount = 7
			}),
			expected: 1,
		},
		{
			name: "only metadata drift (updated) returns 1",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Updated = 1
				s.Unchanged = 9
			}),
			expected: 1,
		},
		{
			name: "empty comparison (total=0) returns 0",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Total = 0
				s.Unchanged = 0
				s.MatchedCount = 0
			}),
			expected: 0,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := ComputeBasicExitCode(tc.summary)
			if got != tc.expected {
				t.Errorf("ComputeBasicExitCode() = %d, want %d", got, tc.expected)
			}
		})
	}
}

// ── ComputeDetailedExitCode ─────────────────────────────────────────────

func TestComputeDetailedExitCode(t *testing.T) {
	tests := []struct {
		name     string
		summary  ComparisonSummary
		expected int
	}{
		{
			name:     "all unchanged returns 0 (identical)",
			summary:  makeSummary(),
			expected: 0,
		},
		{
			name: "fixes only returns 10",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Fixed = 3
				s.Unchanged = 7
			}),
			expected: 10,
		},
		{
			name: "regressions only returns 11",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Regressed = 2
				s.Unchanged = 8
			}),
			expected: 11,
		},
		{
			name: "mixed fixes and regressions returns 12",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Fixed = 2
				s.Regressed = 1
				s.Unchanged = 7
			}),
			expected: 12,
		},
		{
			name: "only new controls returns 13",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.New = 2
				s.Unchanged = 10
				s.Total = 12
				s.UnmatchedNewCount = 2
				s.MatchedCount = 10
			}),
			expected: 13,
		},
		{
			name: "only absent controls returns 13",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Absent = 3
				s.Unchanged = 7
				s.Total = 10
				s.UnmatchedOldCount = 3
				s.MatchedCount = 7
			}),
			expected: 13,
		},
		{
			name: "new and absent but no status changes returns 13",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.New = 1
				s.Absent = 1
				s.Unchanged = 8
				s.Total = 10
				s.UnmatchedNewCount = 1
				s.UnmatchedOldCount = 1
				s.MatchedCount = 8
			}),
			expected: 13,
		},
		{
			name: "only metadata drift (updated) returns 14",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Updated = 1
				s.Unchanged = 9
			}),
			expected: 14,
		},
		{
			name: "fixes and new controls returns 10 (fixes priority over baseline)",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Fixed = 2
				s.New = 1
				s.Unchanged = 7
				s.Total = 10
				s.UnmatchedNewCount = 1
				s.MatchedCount = 9
			}),
			expected: 10,
		},
		{
			name: "regressions and absent controls returns 11 (regressions priority)",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Regressed = 1
				s.Absent = 2
				s.Unchanged = 7
				s.Total = 10
				s.UnmatchedOldCount = 2
				s.MatchedCount = 8
			}),
			expected: 11,
		},
		{
			name: "fixes regressions new absent all present returns 12 (mixed priority)",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Fixed = 1
				s.Regressed = 1
				s.New = 1
				s.Absent = 1
				s.Unchanged = 6
				s.Total = 10
				s.UnmatchedNewCount = 1
				s.UnmatchedOldCount = 1
				s.MatchedCount = 8
			}),
			expected: 12,
		},
		{
			name: "updated and new returns 13 (baseline priority over drift)",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Updated = 1
				s.New = 1
				s.Unchanged = 8
				s.Total = 10
				s.UnmatchedNewCount = 1
				s.MatchedCount = 9
			}),
			expected: 13,
		},
		{
			name: "empty comparison (total=0) returns 0",
			summary: makeSummary(func(s *ComparisonSummary) {
				s.Total = 0
				s.Unchanged = 0
				s.MatchedCount = 0
			}),
			expected: 0,
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := ComputeDetailedExitCode(tc.summary)
			if got != tc.expected {
				t.Errorf("ComputeDetailedExitCode() = %d, want %d", got, tc.expected)
			}
		})
	}
}
